### PR TITLE
INT-4299: Add AbstractMailReceiver.autoCloseFolder

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/INT-4299

* Optimize `AbstractMessageSource` do not evaluate headers when polling
result is `null`
* Optimize `MailReceivingMessageSource` do not wrap polling result to
the message: the `AbstractMessageSource` will do that later

### CODE REVIEW

**Comment by artembilan** [comment]
If this is OK, I'll add some docs on the matter.

Thanks

**Review by garyrussell — CHANGES_REQUESTED** [review]
Just minor javadoc polishing.

LGTM - awaiting docs.

File: spring-integration-mail/src/main/java/org/springframework/integration/mail/AbstractMailReceiver.java [hunk 1]
```diff
@@ -357,7 +378,9 @@ private Folder obtainFolderInstance() throws MessagingException {
 				return convertMessagesIfNecessary(searchAndFilterMessages());
 			}
 			finally {
-				MailTransportUtils.closeFolder(this.folder, this.shouldDeleteMessages);
+				if (this.autoCloseFolder) {
+					closeFolder();
```
Comment (garyrussell) [inline spring-integration-mail/src/main/java/org/springframework/integration/mail/AbstractMailReceiver.java]: Need to close here if no messages were retrieved (needs a test),

Comment (artembilan) [reply spring-integration-mail/src/main/java/org/springframework/integration/mail/AbstractMailReceiver.java]: Why? Why do we need to close the folder at all?
I don't see any comment in the Internet requiring to close the folder after each fetch.
😕

Comment (garyrussell) [reply spring-integration-mail/src/main/java/org/springframework/integration/mail/AbstractMailReceiver.java]: ok

**Review by garyrussell — APPROVED** [review]

**Pull request merged**